### PR TITLE
`CURLOPT_POSTFIELDS` only set when the request body content is not empty

### PR DESCRIPTION
**Pull request by jason-horvath**
I did open an issue on this with an explainer but figured it couldn't hurt to put in a PR. Sorry if it seems spammy.  :)
https://github.com/cakephp/cakephp/issues/13639

### CODE REVIEW

File: src/Http/Client/Adapter/Curl.php [hunk 1]
```diff
@@ -110,9 +110,10 @@ public function buildOptions(RequestInterface $request, array $options): array
         }
 
         $body = $request->getBody();
-        $body->rewind();
-        $out[CURLOPT_POSTFIELDS] = $body->getContents();
-
+        if (!empty($body->getContents())) {
+            $body->rewind();
+            $out[CURLOPT_POSTFIELDS] = $body->getContents();
```
Comment (garas) [inline src/Http/Client/Adapter/Curl.php]: Retrieve (after rewind) contents once, then check `!== ''`.

Your `if` now checks non-rewinded contents and might be at end of stream. Also `empty` is useful on variables, not function calls.

File: src/Http/Client/Adapter/Curl.php [hunk 2]
```diff
@@ -110,9 +110,10 @@ public function buildOptions(RequestInterface $request, array $options): array
         }
 
         $body = $request->getBody();
-        $body->rewind();
-        $out[CURLOPT_POSTFIELDS] = $body->getContents();
```
Comment (markstory) [inline src/Http/Client/Adapter/Curl.php]: Could you unset the field key if the value has no length?

File: src/Http/Client/Adapter/Curl.php (hunk 1 above)
Comment (KainGNX) [reply src/Http/Client/Adapter/Curl.php]: I could make these changes, I see the other proposed change by @markstory , I take it these are two different solutions.

File: src/Http/Client/Adapter/Curl.php (hunk 2 above)
Comment (KainGNX) [reply src/Http/Client/Adapter/Curl.php]: Pushed up a commit that just unsets `CURLOPT_POSTFIELDS` based on this comment.

File: src/Http/Client/Adapter/Curl.php [hunk 3]
```diff
@@ -108,14 +108,14 @@ public function buildOptions(RequestInterface $request, array $options): array
                 $out[CURLOPT_CUSTOMREQUEST] = $request->getMethod();
                 break;
         }
-
```
Comment (KainGNX) [inline src/Http/Client/Adapter/Curl.php]: Thanks @stickler-ci that was fast. I was wondering where these went, I was gonna do the same! :D

**Comment by markstory** [comment]
Thank you 👍

**Comment by dereuromark** [comment]
Do we need to baclport to 3.x?

**Comment by markstory** [comment]
@dereuromark No 3.x didn't have the changes that caused this issue applied as they were part of the psalm/phpstan changes.

**Pull request merged**